dockerfile: QAT: Add an automated QAT kernel/rootfs builder

grahamwhaley opened the pull request:
Add a method to automatically build a modified kernel and rootfs
to support QAT.

Given we don't generally have either:
 - a fully enabled ('full fat') kernel for Kata, as it would be
   big and slow
 - kernel and rootfs variants, as that would seriously bloat our
   test matrix and CI requirements, which would likely lead to more
   strange and hard to find bugs....

maybe we can build a collection of 'helpers' to make building and
customizing a rootfs and kernel easier.

Fixes: #440

Signed-off-by: Graham Whaley <graham.whaley@intel.com>

grahamwhaley commented:
@eadamsintel - PTAL.
If you could test if the resulting binaries actually work, that would be great.
@jcvenegas @chavafg - we should probably discuss if we think this is the right way and the right place to do this.
@jodh-intel , yes, I would have liked to have made the script auto generated from the Kata QAT markdown doc - but, I think that would involve a pretty major rewrite of the doc... which, we could debate on, but then somebody else (:cough: @eadamsintel) would probably have to do.

/cc @mythi fyi

I expect this to need a little rework at least - probably a few extra bits in the docs, and maybe the binaries don't actually work yet?

mythi commented on dockerfiles/QAT/run.sh:
```diff
@@ -0,0 +1,149 @@
+	KERNEL_ROOTFS_DIR=${KERNEL_MAJOR_VERSION}.${KERNEL_PATHLEVEL}.${KERNEL_SUBLEVEL}${KERNEL_EXTRAVERSION}
+	cd $QAT_SRC
+	KERNEL_SOURCE_ROOT=${linux_kernel_path} ./configure ${QAT_COMPILE_OPTIONS}
+	make all -j$(nproc)
```
`make qat-driver-all quickassist-all` saves you from that g++ dependency.

grahamwhaley replied:
nice, but, sadly the 'configure' looks for the C++ requirement, so will fail if I don't have g++ installed... and I'm not guru enough with `configure` to know/see if we can tell it we really don't need g++. If somebody knows, shout - otherwise, I'll let it live as it is.

mythi replied:
I took a quick look and it looks the check cannot be skipped/disabled, so leave it as it is.

grahamwhaley replied:
I did/will change the make line though - every little saving helps ;-) thx!

mythi commented:
@grahamwhaley looks nice to me!

eadamsintel commented:
I tried this on Clear Linux and it failed to build.

```
$ curl -O https://raw.githubusercontent.com/kata-containers/osbuilder/1c34a54bff71c7c644f6f67af505b220f8452127/dockerfiles/QAT/Dockerfile
$ docker build --build-arg http_proxy=http://<proxyinfo> --build-arg https_proxy=<proxyinfo> .
.......skip some ouput......
  pkgconf-1.6.3-2.fc31.x86_64
  pkgconf-m4-1.6.3-2.fc31.noarch
  pkgconf-pkg-config-1.6.3-2.fc31.x86_64
  zlib-devel-1.2.11-19.fc31.x86_64

Complete!
Removing intermediate container e4331565a1d2
 ---> 272168663999
Step 9/11 : COPY ./run.sh /input/
COPY failed: stat /var/lib/docker/tmp/docker-builder467995656/run.sh: no such file or directory
```

It also failed on a Centos system I had in the same way.
```
  perl-threads-shared-1.60-440.fc31.x86_64
  pkgconf-1.6.3-2.fc31.x86_64
  pkgconf-m4-1.6.3-2.fc31.noarch
  pkgconf-pkg-config-1.6.3-2.fc31.x86_64
  qemu-img-2:4.1.1-1.fc31.x86_64
  systemd-devel-243.7-1.fc31.x86_64
  whois-nls-5.5.6-1.fc31.noarch
  xz-5.2.4-6.fc31.x86_64
  zlib-devel-1.2.11-20.fc31.x86_64

Complete!
Removing intermediate container 663e5f104485
 ---> 148c8260f352
Step 9/11 : COPY ./run.sh /input/
COPY failed: stat /var/lib/docker/tmp/docker-builder095040869/run.sh: no such file or directory
```

jodh-intel commented on dockerfiles/QAT/qat.conf:
```diff
@@ -0,0 +1,9 @@
+CONFIG_PCIEAER=y
```
Sort alphabetically? ;)

grahamwhaley replied:
hmmm. I suspect they are currently in the order they either natively turn up in a kernel config file or that is recommended by some QAT docs somewhere, or 'make sense' in the order they are in (kernel kconfig files are mildly ordered). I'm prone to leave as is unless @mythi or @eadamsintel say differently

devimc replied:
wait, is this a kernel fragment? why it's not in the fragments directory (like the gpu fragment)?

grahamwhaley replied:
because.... if you read the intent and the code - this gets copied over to the fragment dir during the build. We don't have QAT enabled in kata kernels by default - half the point of this PR is to add a semi-automated way to do a custom Kata rootfs/kernel build ;-)

grahamwhaley replied:
but, I guess I understand maybe what you mean - 'Why do we not just enable this in the default kernel?'. I guess size and stuff. This dockerfile is not different from the current QAT setup we have, which is a markdown doc describing how to do this. We didn't put the QAT kernel fragment in the default kernel at that point either. the current decision I think is to not have all possible hardware enabled in the default kernel. Each separate item I guess requires its own discussion on its merits. As this PR does not change the current status quo, I'll leave it as is, and if somebody wants to debate if QAT should be enabled in the default kata kernel, then raise the question here, or a new Issue, or in the weekly conf call I guess?

jodh-intel commented on dockerfiles/QAT/qat.conf:
```diff
@@ -0,0 +1,9 @@
+CONFIG_PCIEAER=y
+CONFIG_UIO=y
+CONFIG_CRYPTO_HW=y
+CONFIG_CRYPTO_DEV_QAT_C62XVF=m
```
Do we mean a module?

grahamwhaley replied:
@mythi @eadamsintel - do you have a definitive answer? I suspect the answer is 'yes' for some reason or other (I wanted to say 'out of tree driver', but this is of course the kernel config...)

mythi replied:
yes, a module. it's a trick: by setting `=m` Kconfig enables the rest of the QAT dependencies we need. it cannot be `y=` because we don't want that driver but the out-of-tree one we build to rootfs later.

grahamwhaley replied:
heh, I'll see if I can add a comment to that effect - I think I can put # comments in the fragments..... as long as they don't explicitly mention `CONFIG_...` names (as the magic tool greps and breaks at that point iirc).

jodh-intel commented on dockerfiles/QAT/run.sh:
```diff
@@ -0,0 +1,149 @@
+}
+
+help() {
+	usage=$(cat << EOF
```
Nit you could just do:


```sh
cat <<EOT

  blah, blah

EOT
```

... to avoid creating the variable.

grahamwhaley replied:
done

jodh-intel commented on dockerfiles/QAT/README.md:
````diff
@@ -0,0 +1,59 @@
+The process is executed using Docker:
+
+```sh
+$ cd ..
````
Where are we? :smile: Can you `cd` to a known directory as it's unclear what directory the user is/should be in at this point.

jodh-intel replied:
Well this is highly unusual I think. Maybe if I turn the question round: why do we need to change directory? I thought idocker would build an image if it found `Dockerfile` in the current directory, so why leave it?

jodh-intel commented on dockerfiles/QAT/run.sh:
```diff
@@ -0,0 +1,149 @@
+set -e
+
+# NOTE: Some env variables are set in the Dockerfile - those that are
+# intended to be over-rideable.
```
Might be worth adding `set -o nounset` to avoid surprises if the `Dockerfile` changes?

Also, how about adding the standard idiom to make debugging easier:

```sh
[ -n "${DEBUG:-}" ] && set -o xtrace
```

grahamwhaley replied:
`set -u` added.
There is already `-d` debug option on the command line - will that suffice?

devimc replied:
and might be worth commenting that this file is intended to run in a container, not in the host

grahamwhaley replied:
See other comment - it is in a subdir called `dockerfiles`, next to a file called `Dockerfile`..... seems fairly self explanatory...no?

jodh-intel commented on dockerfiles/QAT/run.sh:
```diff
@@ -0,0 +1,149 @@
+runtimerepo=github.com/kata-containers/runtime
+runtimerepopath=${GOPATH}/src/${runtimerepo}
+agentrepo=github.com/kata-containers/agent
+agentrepopath=${GOPATH}/src/${agentrepo}
```
This is a little hard to read - maybe space into blocks?

grahamwhaley replied:
Sure. had considered it. swings and roundabouts.

jodh-intel commented on dockerfiles/QAT/run.sh:
```diff
@@ -0,0 +1,149 @@
+
+configure_kernel()
+{
+	cp /input/qat.conf ${packagerepopath}/kernel/configs/fragments/common/qat.conf
```
Seeing this absolute path makes me think it might be worth adding a brief description at the top of the file stating that this is *only* designed to run inside a container.

grahamwhaley replied:
well, it is sat in a folder called `dockerfiles` alongside a `Dockerfile`.... I think it's fairly self explanatory...

jodh-intel replied:
I think you may underestimate the power of people to make mistakes :smile: It's a script, so somebody is *bound* to try to run it on their system and it will fail because it won't be in a container environment.

If there was a realiable way to detect you were in a docker env, I'd add a test at the top of the script to assert that personally.

grahamwhaley commented:
> I tried this on Clear Linux and it failed to build.

OOI @eadamsintel - what is your docker default runtime? I suspect it might be Kata, and 'docker build' is an issue with Kata as the default runtime iirc. You also iirc cannot use `--runtime=runc` on a `docker build` - so, you might have to switch Docker default runtimes. Ultimately, it might come down to a classic Kata 9p issue ... Let me know. I used Clear as the host when I wrote/tested this - but with runc as the default runtime.

devimc commented on dockerfiles/QAT/README.md:
````diff
@@ -0,0 +1,59 @@
+$ docker run -ti --rm --privileged -v /dev:/dev -v $(pwd)/output:/output kataqat
+```
+
+> **Note:** The use of the `--privileged` and `-v /dev:/dev` arguments to the `docker run` are
````
so, the will not work in a kata container?

grahamwhaley replied:
probably correct - just like osbuilder will not afaik - it does the same thing in `USE_DOCKER` mode I think

devimc commented on dockerfiles/QAT/README.md:
```diff
@@ -0,0 +1,59 @@
+| KATA_VERSION | Kata Branch or Tag to build from | master |
+| OUTPUT_DIR | Directory inside container where results are stored | /output |
+
+Variables can be set on the `docker run` commandline, for example:
```
podman is also supported, isn't it?

grahamwhaley replied:
No idea. I'm not set up to run it. It is a **Dockerfile** ;-) - if somebody wants to try/confirm that this also works with other container builder/managers, I or they can add it to the doc...

eadamsintel commented:
> what is your docker default runtime? I suspect it might be Kata,

It is actually runc in both cases according to `docker info`

eadamsintel commented on dockerfiles/QAT/run.sh:
```diff
@@ -0,0 +1,158 @@
+{
+	# Check out all the repos we will use now, so we can try and ensure they use the specified branch
+	# Only check out the branch needed, and make it shallow and thus space/bandwidth efficient
+	git clone --single-branch --branch $KATA_VERSION --depth=1 https://${packagerepo} ${packagerepopath}
```
This failed for me because it did not pick up my proxy information.

grahamwhaley replied:
Can you show the exact error you got? I want to check if that was git'ing via https or the git: protocol. It should afaik be using the https one, in which case it should pick up afaik your docker (host) configured proxy info.

eadamsintel replied:
Both my $http_proxy and $https_proxy variables are set. 

```
$ docker run -ti --rm --privileged -v /dev:/dev -v $(pwd)/output:/output kataqat
Cloning into '/root/src/go/src/github.com/kata-containers/packaging'...
fatal: unable to access 'https://github.com/kata-containers/packaging/': Failed to connect to github.com port 443: Connection timed out
```

If I git clone on the host it works without issue.  In my next comment I suggested to clone as part of the Dockerfile. That might help with this issue as well. You are doing the clone from a script inside a docker container vs on the Docker ADD line as part of building the container. Someone can successfully build a Docker image but then that image won't connect to anything because of proxies issues and it will be harder to debug. However, if the clones are done as part of the ADD command and they fail there then the container image won't get built and it is more obvious it is a proxy connection issue that can be fixed by passing the http_proxy and https_proxy variables on the `docker build` command.

grahamwhaley replied:
> Both my $http_proxy and $https_proxy variables are set.

Set where - in your host env, or in your docker config?

eadamsintel replied:
It is set in both.

```
$ echo $http_proxy
http://proxy:port/
$ echo $https_proxy
http://proxy:port//


$ cat /etc/systemd/system/docker.service.d/proxy.conf
[Service]
Environment="HTTP_PROXY=http://proxy:port/"
Environment="HTTPS_PROXY=http://proxy:port/"
```

grahamwhaley replied:
Ah, the bit you are maybe missing is to get the proxy settings *inside* the containers. The above I believe sets them:
- on the host
- in the systemd unit of the docker server

I then placed this in my `~/.docker/config.json`:

```json
{
 "proxies":
 {
   "default":
   {
     "httpProxy": "http://proxy:port",
     "httpsProxy": "http://proxy:port"
   }
 }
}
```

which gets the proxy information *into* the containers themselves, and then the image ran and built for me. You want to try that @eadamsintel ?

mythi replied:
> I then placed this in my `~/.docker/config.json`

This is the only sane setting when working with proxies.

eadamsintel replied:
I put the proxies in both locations and can build and run containers successfully.

eadamsintel commented on dockerfiles/QAT/Dockerfile:
```diff
@@ -0,0 +1,51 @@
+    xz
+
+# Pull in our local files
+COPY ./run.sh /input/
```
I had to manually `chmod +x` this before building the container image when I downloaded the file from the GitHub raw using curl.

grahamwhaley replied:
that's because your `curl` dropped the execute bit from the script. Don't do that - that is not how you get code from github ;-) - you need to `git clone` the repo :-)

eadamsintel replied:
Yes it took me a bit of time to figure that out. This is the first time I had run into this particular issue because I always curl Dockerfile's from github directly. Maybe that is the wrong way to do it, but I have seen instructions throughout the years that do just that.  There are others who are like me that do things perhaps a non-standard way and we have to design for that.  That's why I am here to test :)    Couldn't you just git clone this osbuilder repo as part of the Dockerfile and then add in the additional files then?
```
RUN git clone <osbuilder repo>
```
That would eliminate the need to download multiple files to the same folder, would eliminate the possibility someone curl's a file and loses the execute permission, and would allow the Dockerfile to just be emailed to people and work. It might be a little tricky to test before committing all the files, but it would make things easier on the user.  You could just email the Dockerfile around and that would be all you need. Also, instead of trying to put all the complex logic of `run.sh` into a Dockerfile to eliminate the need of that file you could just clone all the necessary scripts needed to build the QAT container.

grahamwhaley replied:
I'm going to say nack to the idea of 'just one dockerfile', for a number of reasons:
 - nobody is going to `curl` down just the dockerfile and run it - **unless** we tell them to do that - and, we don't. They will only find the Dockerfile and docs by visiting the repo and reading the instructions, which do not say `curl` down one file and run it... I'll make it even clearer in the documentation you need to git clone (or at least have the whole QAT directory) present. You happen to have curl'd the file down on its own because you only read a fragment of the PR. once the PR is merged, nobody will be doing that, they will start from the markdown instructions.
 - wiring those git urls into the Dockerfile pins the dockerfile to precisely that version (master) of what is on github, and to that path (url). **not** wiring those into the Dockerfile means the `docker build` will work wherever you place the files (no hard wired paths), and it allows you to edit your local version and build it. Both good things. Let's not wire hard paths into the Dockerfile.
 - putting that git clone into the Dockerfile, as above, would mean the git repo would end up in the docker image as well. There are ways to mitigate that (both with a shallow clone and 're-basing' the Dockerfile after the clone and copy), but, really, this is all adding un-necessary complexity. As it stands, the Dockerfile and additional COPY files fit pretty well with the KISS principle and are standard Dockerfile practices. Let's keep it that way.

eadamsintel replied:
Fair enough. My goal is to make this as easy as possible for end users even if it is a bit more complex to maintain.

eadamsintel commented:
I built this on a different system without proxy and the kernel and rootfs did work. However, I had to find a qat configuration file with the [shim] section added to use with the container. The manual instructions takes the qat configuration file from the driver build, edits a section to make it a [shim] section and then uses that as part of the container run.  I think this file should be provided in the /output folder as well but  @mythi  might have a different opinion.

mythi commented:
> /output folder as well but @mythi might have a different opinion.

a sample config probably makes sense especially if we use the artifacts in some CI one day. normally, the workload would bring it's own file to be able to configure the device based on the workload needs (e.g., to make the device crypto only)

grahamwhaley commented:
> > /output folder as well but @mythi might have a different opinion.
> 
> a sample config probably makes sense especially if we use the artifacts in some CI one day. normally, the workload would bring it's own file to be able to configure the device based on the workload needs (e.g., to make the device crypto only)

ack - looking at the existing Kata QAT doc, we copy one out of the QAT source - I'll take inspiration from that, likely make a `configs` subdir in the output dir, and copy them there - and will add notes to the README.
Once this is all done, I'll go edit the existing Kata QAT doc to then start to reference this dockerfile.

grahamwhaley commented:
@eadamsintel @mythi - I've re-pushed with the addition of copying the example QAT `vm.conf` files out to the `output/configs` directory. ptal

eadamsintel commented:
> I've re-pushed

I tested on a Clearlinux and Centos system and both worked with using their default QAT host driver. I had to edit the `c6xxvf_dev0.conf.vm` file to add a `[SHIM]` section.  Actually, I just changed the `[SSL]` section to be `[SHIM]`.  On my docker command line I changed it to be `-v /home/eadams/osbuilder/dockerfiles/QAT/output/configs/c6xxvf_dev0.conf.vm:/etc/c6xxvf_dev0.conf` instead of passing in a folder and mapping to `/etc`     

If you don't do this then after a successful `adf_ctl restart` in the Kata container you will get the following error when running the following commands. The error is not very clear so this should be added to the help section of the QAT guide. 

**No SHIM section error message**
```
root@cf313ae050fb/ # adf_ctl restart
Restarting all devices.
Processing /etc/c6xxvf_dev0.conf
root@cf313ae050fb/ # openssl engine -c -t qat
(qat) Reference implementation of QAT crypto engine
 [RSA, DSA, DH, AES-128-CBC-HMAC-SHA1, AES-128-CBC-HMAC-SHA256, AES-256-CBC-HMAC-SHA1, AES-256-CBC-HMAC-SHA256, TLS1-PRF, HKDF]
     ADF_UIO_PROXY err: icp_adf_userProcessToStart: Error reading /dev/qat_dev_processes file
[ unavailable ]
```

Since the `run.sh` script is embedded in the Docker container and someone might just `curl` the file instead of doing a `git clone` I think before the `run.sh` command is executed if it is possible the user should be told that this script needs to be executable through `chmod`  That way if it fails at that step they will see the message about `run.sh` giving clues what to do next. I am not sure if that can be embedded in a Dockerfile.  Alternatively, consider embedding the `git clone` statement in the `Dockerfile` so that it grabs all the necessary files as part of building the Dockerfile. This eliminates that mistake that someone might just download only the Dockerfile to build. Also, it makes it easy to just email the Dockerfile around and still be buildable. 

I also think it would be useful in the `run.sh` before the first command that downloads any code to `echo` to the user a warning that if the connection fails and the user is behind a proxy that their proxy is properly set. This way if it fails at that step the last thing users will see is the proxy warning giving them clues on how to fix.

eadamsintel commented:
After talking with @mythi I found that you can use the config file as is if you just specify a `QAT_SECTION_NAME` variable and make it equal to one of the existing sections in the conf file. In this case set it to the existing SSL section.

```
$ docker run -it --runtime=kata-runtime --cap-add=IPC_LOCK --cap-add=SYS_ADMIN --device=/dev/vfio/90 -v /dev:/dev -v /home/eadams/osbuilder/dockerfiles/QAT/output/configs/c6xxvf_dev0.conf.vm:/etc/c6xxvf_dev0.conf openssl-qat-engine bash
root@a4bc9835dc40/ # adf_ctl restart
Restarting all devices.
Processing /etc/c6xxvf_dev0.conf
root@a4bc9835dc40/ # QAT_SECTION_NAME=SSL openssl engine -c -t qat
(qat) Reference implementation of QAT crypto engine
 [RSA, DSA, DH, AES-128-CBC-HMAC-SHA1, AES-128-CBC-HMAC-SHA256, AES-256-CBC-HMAC-SHA1, AES-256-CBC-HMAC-SHA256, TLS1-PRF]
     [ available ]
```

grahamwhaley commented:
> Since the `run.sh` script is embedded in the Docker container and someone might just `curl` the file instead of doing a `git clone` I think before the `run.sh` command is executed if it is possible the 

Really, I think you need to ignore the case where 'somebody might `curl` a file' - I don't think it's going to happen once the PR is merged. I think you are trying to fix a case that does not exist, and making things much more complex than they need to be.

I appreciate the idea of 'can email the Dockerfile around', and if a concensus is that we need to support that mode of operation (which I have to state, is *not* a standard Dockerfile process), then I would suggest we have to embed the `run.sh` and `qat.conf` files inside the Dockerfile with shell magic `cat <<` operations, unclean as they may be.

grahamwhaley commented:
Updated. I've elaborated the fact the config files will need modifying, and pointed back to the use-case QAT doc for more info.
I also noted that you need the whole QAT directory to build the Dockerfile (but as we know, if you don't have it, it won't build anyway...)
I've also modified the top level doc in this repo to reference this one and briefly mention customized images - mostly to keep the doc check CI happy ;-)
I've dropped the RFC, so let's see what the rest of the CI thinks now.

grahamwhaley commented on dockerfiles/QAT/README.md:
```diff
@@ -0,0 +1,71 @@
+
+## Building
+
+The process is executed using Docker, from the directory *above* this `QAT` folder. You
```
Ack. have reworded.

grahamwhaley commented on dockerfiles/QAT/README.md:
```diff
@@ -0,0 +1,71 @@
+When complete, the generated files will be placed into the output directory. Example QAT config
+files are also placed into the `config` subdirectory for references. These files may need
+modification before use, as documented in the Kata QAT
+[use-case document](https://github.com/kata-containers/documentation/blob/master/use-cases/using-Intel-QAT-and-kata.md#copy-intel-qat-configuration-files-and-enable-virtual-functions):
```
One of the links has a trailing section `#` link (['fragment'?](https://github.github.com/gfm/#example-509)). I don't know if you can do that by concatenating onto a [reference link](https://github.github.com/gfm/#reference-link) ?
(and, yes, I had to go look up what a 'reference link' was, as I pretty much never ever put those into markdown docs :-) ).

wdyt - if it can be done, I can add it - but, I don't know if/how to do that.... with the links being subtly different...

jodh-intel replied:
Good point - let's leave as-is then.

grahamwhaley commented on dockerfiles/QAT/README.md:
````diff
@@ -0,0 +1,71 @@
+Variables can be set on the `docker run` commandline, for example:
+
+```sh
+docker run -ti --rm --privileged -e "KATA_VERSION=1.11.0" -v /dev:/dev -v $(pwd)/output:/output kataqat
````
`$` added.
Heh, you know, I don't think I've ever used `${PWD}` - sub'd in for you ;-)
For you entertainment then, try peeking at [this](https://github.com/search?l=Shell&q=org%3Akata-containers+%24%28pwd%29&type=Code) :-)

grahamwhaley commented:
Hmm, and I'll sneak a peek at the CI spell mumbles....

```
INFO: Spell checking file 'dockerfiles/QAT/README.md'
WARNING: Word 'gz': did you mean one of the following?: g, z, gs, gr, gt, go, oz, dz, gm, Oz, Hz
WARNING: Word 'icp': did you mean one of the following?: pic, ice, imp, icy
WARNING: Word 'lkcf': did you mean one of the following?: folk
WARNING: Word 'qat': did you mean one of the following?: qt, at, eat, sat, rat, tat, oat, lat, cat, mat, pat, hat, bat, fat, vat
WARNING: Word 'sriov': did you mean one of the following?: Rios
ERROR: Spell check failed for file: 'dockerfiles/QAT/README.md'
```

jodh-intel commented on dockerfiles/QAT/Dockerfile:
```diff
@@ -0,0 +1,51 @@
+# Kata osbuilder 'works best' on Fedora
+FROM fedora:latest
+
+# Version of the Dockerfile
```
Might be worth a comment here explaining when this should be changed?

jodh-intel commented on dockerfiles/QAT/README.md:
```diff
@@ -0,0 +1,71 @@
+
+| Variable | Definition | Default value |
+| -------- | ---------- | ------------- |
+| QAT_DRIVER_VER | QAT driver version to use | `qat1.7.l.4.9.0-00008.tar.gz` |
```
Nit: Sort rows? :smile:

jodh-intel commented on dockerfiles/QAT/run.sh:
```diff
@@ -0,0 +1,160 @@
+{
+	local check_in_container=${OUTPUT_DIR:-}
+	if [ -z "${check_in_container}" ]; then
+		echo "Error: 'OUTPUT_DIR' not set"
```
Nit: Redirect error message to `stderr`?

grahamwhaley commented:
/test
Added a copyright header to the `qat.conf` to keep the CI happy.

jodh-intel commented:
We have green, but maybe we should get a final ack from @eadamsintel or @mythi on this one.

grahamwhaley commented:
<cough> green...

grahamwhaley commented:
@jodh-intel ack, ideally both of them.

jodh-intel commented:
Added the dnm to prevent any over eager green button pouncers (myself included...)

eadamsintel commented:
I just rebuilt this which used the slightly newer `vmlinuz-kata-linux-5.4.32-75_qat` and everything worked fine.  It looks good to me.

jodh-intel commented:
Any last minute comments @mythi ?

mythi commented:
@jodh-intel sorry for the delay. this is great and good to merge from my side!

@eadamsintel have you tested the 4.9 release?

@grahamwhaley thanks!

jodh-intel commented:
Thanks @mythi. Once @eadamsintel has confirmed the kernel version, we can remove the dnm label and merge now :smile:

eadamsintel commented:
@mythi Yes I tested the 4.9 release.  The Dockerfile has `ENV QAT_DRIVER_VER qat1.7.l.4.9.0-00008.tar.gz`

jodh-intel commented:
Right - label removed. I'm not sure why I'm the only one showing as having approved this PR (anyone else... ?)

jodh-intel commented:
Thanks all - let's do this...!
